### PR TITLE
Fix known token contract address warning

### LINKED ISSUE #12642: Warning not being shown when sending tokens to a known token contract address
In `ducks/send/send.js` we have code that is meant to show a warning to users when they are sending to a known contract address:

```
if (
          isSendingToken &&
          isValidHexAddress(recipient.userInput) &&
          (tokenAddressList.find((address) =>
            isEqualCaseInsensitive(address, recipient.userInput),
          ) ||
            checkExistingAddresses(recipient.userInput, tokens))
        ) {
          recipient.warning = KNOWN_RECIPIENT_ADDRESS_WARNING;
        } else {
          recipient.warning = null;
 }
```

This `recipient.warning` property is read in `renderDialogs` of `ui/pages/send/send-content/add-recipient/add-recipient.component.js`

Currently, if I try to send DAI token to the DAI contract, no warning is shown.

Acceptance criteria:
- if a known token contract address is entered in the recipient field in the send flow, a warning should be shown to the user
- this warning can just be the `<Dialog type="warning" className="send__error-dialog">` that `add-recipient.component.js` is currently meant to render
- "known token contract addresses" include any of the addresses in this list: https://github.com/MetaMask/contract-metadata/blob/master/contract-map.json

### PR DESCRIPTION
**Pull request by filipsekulic**
Fixes: https://github.com/MetaMask/metamask-extension/issues/12642

**Screencast**:

https://user-images.githubusercontent.com/92310504/152987260-6dee5ed3-9d53-4b62-8c60-6bf17909695a.mov

### CODE REVIEW

**Comment by mirjanaKukic** [comment]
Verified by QA.

**Review by david0xd — DISMISSED** [review]

File: ui/pages/send/send-content/send-content.component.js
Comment (adonesky1) [inline ui/pages/send/send-content/send-content.component.js:23]: I think for clarity we should call this `isKnownContractAddress`

File: ui/pages/send/send-content/send-content.component.js
Comment (adonesky1) [inline ui/pages/send/send-content/send-content.component.js:97]: do we require both conditions here? Wouldn't  `result.symbol !== undefined || result.decimals !== undefined` be sufficient to establish that its on the list and what we would consider a know token/contract address?

File: ui/pages/send/send-content/send-content.component.js
Comment (adonesky1) [inline ui/pages/send/send-content/send-content.component.js:111]: any reason this shouldn't just be called in the `render` method?

File: ui/pages/send/send-content/send-content.component.js
Comment (adonesky1) [inline ui/pages/send/send-content/send-content.component.js:126]: Wondering if we should rename this key to `knownContractAddressWarning`? cc @danjm

**Review by adonesky1 — COMMENTED** [review]
A few small questions

File: ui/pages/send/send-content/send-content.component.js
Comment (filipsekulic) [reply ui/pages/send/send-content/send-content.component.js]: Fixed in: ad39793344c75085bf0a75435319383fa7f7b514

File: ui/pages/send/send-content/send-content.component.js
Comment (filipsekulic) [reply ui/pages/send/send-content/send-content.component.js]: You have got the point.
Fixed in: ad39793344c75085bf0a75435319383fa7f7b514

File: ui/pages/send/send-content/send-content.component.js
Comment (filipsekulic) [reply ui/pages/send/send-content/send-content.component.js]: Fixed in: ad39793344c75085bf0a75435319383fa7f7b514

File: ui/pages/send/send-content/send-content.component.js
Comment (filipsekulic) [reply ui/pages/send/send-content/send-content.component.js:124]: Did not change anything. Waiting for Dan's response on this one.

**Review by adonesky1 — DISMISSED** [review]
LGTM. Still think we should change that key name but won't block on it.

File: ui/ducks/send/send.js
Comment (darkwing) [inline ui/ducks/send/send.js:997]: Can this be `asset?.details?.address ?? ''`?

File: ui/pages/send/send-content/send-content.component.js
Comment (darkwing) [inline ui/pages/send/send-content/send-content.component.js:95]: Can we make this method async so we can use `async/await` instead of `then`?

File: ui/ducks/send/send.js
Comment (filipsekulic) [reply ui/ducks/send/send.js]: Yes, it can be. 
Fixed in: 51054012cfbb3e51b2853adafa9e9d8fd155656d

File: ui/pages/send/send-content/send-content.component.js
Comment (filipsekulic) [reply ui/pages/send/send-content/send-content.component.js]: Yes, we can. I changed it.
Fixed in: 51054012cfbb3e51b2853adafa9e9d8fd155656d

File: ui/pages/send/send-content/send-content.component.js
Comment (filipsekulic) [reply ui/pages/send/send-content/send-content.component.js]: Actually, I had to modify this condition. The reason for doing so is the fact that entering some user's address gives the result: 
{ 
  symbol: '',
  decimals: '0'
}.
If the condition was just `result.symbol !== undefined || result.decimals !== undefined`, it would pass. 
Concerning that, I added another check, so the final condition looks like `if ((symbolAndDecimals.symbol !== undefined || symbolAndDecimals.decimals !== undefined) && symbolAndDecimals.symbol !== '')`.
If the address is contract, it will have the symbol as well.

File: ui/pages/send/send-content/send-content.component.js
Comment (danjm) [reply ui/pages/send/send-content/send-content.component.js:131]: Let's not change it right now, because then we have to update all of the translations too

**Review by danjm — DISMISSED** [review]
LGTM

File: ui/ducks/send/send.js
Comment (filipsekulic) [inline ui/ducks/send/send.js:1059]: @adonesky1 
I removed this piece of code as the `isSendingToken` is not necessary here for my PR. In order to make sending ETH possible,`|| asset.type ===  ASSET_TYPES.NATIVE` should be added or `isSendingToken` removed from here and down there. 
ASSET_TYPES looks like this:
`export const ASSET_TYPES = {
  NATIVE: 'NATIVE',
  TOKEN: 'TOKEN',
  COLLECTIBLE: 'COLLECTIBLE',
};
`
In this case, if another condition was added, `isSendingToken` would always be true.  
Agreed with Dan about this one.
// cc @danjm

Comment (adonesky1) [reply ui/ducks/send/send.js:1059]: Looks good to me!

Comment (adonesky1) [reply ui/ducks/send/send.js:1059]: Thanks for pointing this out though!

File: ui/pages/send/send-content/send-content.component.js
Comment (adonesky1) [inline ui/pages/send/send-content/send-content.component.js:118]: These knownContract addresses are only on Mainnet so should we only show this warning when we're on Ethereum mainnet? cc @danjm

File: ui/pages/send/send-content/send-content.component.js
Comment (adonesky1) [inline ui/pages/send/send-content/send-content.component.js:97]: Once controllers v25.0.0 is added to the extension we'll have a way check whether the address is an ERC721 or ERC1155 contract as well. Should we wait and add that functionality? I'm happy to pair program and make that change as well. cc @danjm @david0xd

**Review by david0xd — DISMISSED** [review]

File: ui/pages/send/send-content/send-content.component.js
Comment (danjm) [inline ui/pages/send/send-content/send-content.component.js:73]: This should be done in a `componentDidUpdate` call. Ideally, the render method does not have state management side effects. Also, with the `componentDidUpdate` method, we can compare `prevProps.recipient` to `this.props.recipient` and only call `this.checkContractAddress` when they are different. This ensures we don't call it more than necessary, and that we do call it if the recipient changes.

File: ui/pages/send/send-content/send-content.component.js
Comment (danjm) [reply ui/pages/send/send-content/send-content.component.js:100]: I think that can be a separate PR after this PR merges. We should create a separate github ticket for that:

- [ ] Create github ticket for ensuring the known contract address warning in the send flow also apples to NFT tokens

File: ui/pages/send/send-content/send-content.component.js
Comment (danjm) [inline ui/pages/send/send-content/send-content.component.js:100]: @filipsekulic Rereading this now, I realize a made a mistake in previous advice I gave you. We shouldn't call `getSymbolAndDecimals` because that will return symbol and decimals for all contracts that have symbol and decimals APIs implemented, but we only want to show the warning when it is from our token list.

I think if you use the `getTokenMetadata` from the same file, instead of `getSymbolAndDecimals` this will work correctly. With that change, you will also just need to add an optional chaining operator below (e.g. `symbolAndDecimals?.symbol`)

File: ui/pages/send/send-content/send-content.component.js
Comment (danjm) [reply ui/pages/send/send-content/send-content.component.js:121]: No, I think it is okay. Because if we are not on mainnet, then we definitely don't want the user sending to those token addresses. I think improve warnings for that case are being developed, but this actually has benefit in the meantime.

**Review by danjm — CHANGES_REQUESTED** [review]
@filipsekulic Sorry, I checked this again and noticed a couple of other changes that are needed.

File: ui/pages/send/send-content/send-content.component.js
Comment (filipsekulic) [reply ui/pages/send/send-content/send-content.component.js]: Fixed in: e3b7ee44fa5b49dbce8251dcb856212b1b34eb85

File: ui/pages/send/send-content/send-content.component.js
Comment (filipsekulic) [reply ui/pages/send/send-content/send-content.component.js]: Fixed in: e3b7ee44fa5b49dbce8251dcb856212b1b34eb85

**Comment by filipsekulic** [comment]
> @filipsekulic Sorry, I checked this again and noticed a couple of other changes that are needed.

@danjm No problem, it's okay. I fixed it like we agreed (e3b7ee44fa5b49dbce8251dcb856212b1b34eb85).

**Comment by mirjanaKukic** [comment]
Verified by QA

**Review by david0xd — DISMISSED** [review]

**Comment by cyanlink** [comment]
Sorry for asking too much but IMO the "warning" is still not enough, at least you should say "WARNING: known contract address, Read More.". Then in the read more you redirect to [this zendesk page](https://metamask.zendesk.com/hc/en-us/articles/360020028092-What-is-the-known-contract-address-warning-) or maybe a newer verbose explanation, like my proposal at #13486 . I really do not know how to contact the metamask team in the most efficient way (the community post and new request ticket does not draw any attention).

**Comment by bschorchit** [comment]
I believe this PR doesn't include the acceptance criteria that @EHaracic [mentioned here](https://github.com/MetaMask/metamask-extension/issues/12642#issuecomment-970235647): "Remove "New address detected ..." when this specific warning is shown"

I'm saying that by looking at the video only, so I might be wrong or this might have been already fixed later.

**Comment by filipsekulic** [comment]
> I believe this PR doesn't include the acceptance criteria that @EHaracic [mentioned here](https://github.com/MetaMask/metamask-extension/issues/12642#issuecomment-970235647): "Remove "New address detected ..." when this specific warning is shown"
> 
> I'm saying that by looking at the video only, so I might be wrong or this might have been already fixed later.

@bschorchit You are right. Somehow I forgot to add that AC to Dan's AC. Good catch. Thanks!

**Comment by danjm** [comment]
Nice work here @filipsekulic

Some new requirements have come in and there are some edge case that I realized are only fixable if we modify the send duck more. Also, the advice I gave before on getting the symbols and decimals wasn't correct for all cases. Instead of making you go back on this again, I started making changes on top your commit and opened a new PR. So we can close this and I will finish off the work their.

https://github.com/MetaMask/metamask-extension/pull/13588

**Pull request closed without merging**
